### PR TITLE
[docs] Add more details about bootstrapping process to a readme 

### PR DESCRIPTION
**Pull request by vpranckaitis**
**What this PR does / why we need it**:

Adds more details about bootstrapping process to a readme file

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

**Comment by vpranckaitis** [comment]
@notbdu please check for inaccuracies (I don't yet have a good picture of all of this) or if there's anything important to add

cc: @gediminasgu @linasm @soundvibe

File: src/dbnode/storage/bootstrap/bootstrapper/README.md [hunk 1]
```diff
@@ -2,12 +2,78 @@
...
 - `peers`: The peers bootstrapper, used to bootstrap any remaining data from peers. This is used for a full node join too.
 - `commitlog`: The commit log bootstrapper, currently only used in the case that peers bootstrapping fails. Once the current block is being snapshotted frequently to disk it might be faster and make more sense to not actively use the peers bootstrapper and just use a combination of the filesystem bootstrapper and the minimal time range required from the commit log bootstrapper.
     - *NOTE*: the commitlog bootstrapper is special cased in that it runs for the *entire* bootstrappable range per shard whereas other bootstrappers fill in the unfulfilled gaps as bootstrapping progresses.
+- `uninitialized`: The uninitialized bootstrapper, used to bootstrap a node when the whole cluster is new and there are no peers to fetch the data from.
+
+The bootstrappers satisfy the `bootstrap.Source` interface. `AvailableData()` and `AvailableIndex()` methods take shard time ranges that need to be bootstrapped and returns which ranges (possibly a subset) can be fulfilled by this bootstrapper. `Read()` method does the bootstrapping of the available data and index ranges.
```
Comment (linasm) [inline src/dbnode/storage/bootstrap/bootstrapper/README.md:52]:
```suggestion
The bootstrappers satisfy the `bootstrap.Source` interface. `AvailableData()` and `AvailableIndex()` methods take shard time ranges that need to be bootstrapped and return which ranges (possibly a subset) can be fulfilled by this bootstrapper. `Read()` method does the bootstrapping of the available data and index ranges.
```

File: src/dbnode/storage/bootstrap/bootstrapper/README.md [hunk 2]
```diff
@@ -2,12 +2,78 @@
...
+### Peer
+
+- `AvailableData()` - inspects the cluster topology and returns the whole requested time range for shards which have enough available replicas to satisfy consistency requirements
+- `AvailableIndex()` - exactly the same as `AvailableData()`
+- `Read()` - fetches shard data from peers and either persists it or checks-out into memory. Then it builds index segments and either flushes them to disk or keeps in memory
+
+### Uninitialized
+
+- `AvailableData()` - for each shard, inspects its status across peers. If the number of `Initializing` replicas is higher than `Leaving`, the shard is deemed _new_ and available to be bootstrapped by this bootstrapper
+- `AvailableIndex()` - exactly the same as `AvailableData()`
```
Comment (linasm) [inline src/dbnode/storage/bootstrap/bootstrapper/README.md:75]: This statement is the same for every bootstrapper, perhaps we should remove `AvailableIndex` (at least from the readme) as documenting it seems to carry no value?

File: src/dbnode/storage/bootstrap/bootstrapper/README.md [hunk 3]
```diff
@@ -2,12 +2,78 @@
...
+
+- `AvailableData()` - checks which shards have ever reached availability (i.e. is in `Available` or `Leaving` state) and returns the whole requested time range for those shards
+- `AvailableIndex()` - exactly the same as `AvailableData()`
+- `Read()` - for each shard reads the most recent snapshot file, and then reads commit log and checks-out all the series belonging to the namespaces that are being bootstrapped
+
+### Peer
+
+- `AvailableData()` - inspects the cluster topology and returns the whole requested time range for shards which have enough available replicas to satisfy consistency requirements
+- `AvailableIndex()` - exactly the same as `AvailableData()`
+- `Read()` - fetches shard data from peers and either persists it or checks-out into memory. Then it builds index segments and either flushes them to disk or keeps in memory
```
Comment (linasm) [inline src/dbnode/storage/bootstrap/bootstrapper/README.md:70]: Worth to also note the checksum validation?

**Review by soundvibe — COMMENTED** [review]
Maybe it's worth mentioning the most important configuration parameters and how they affect bootstrapping?

File: src/dbnode/storage/bootstrap/bootstrapper/README.md [hunk 4]
```diff
@@ -2,12 +2,74 @@
...
+If there are any errors or unfulfilled time ranges after going through all the configured bootstrappers, the process is retried. 
+
+See pseudo-code below for more details.
+
+```python
+bootstrappers = {filesystem, commitlog, peer, uninitialized}
+
+while true:
+    namespaces = getOwnedNamespaces()
+    shards = filterNonBootstrapped(getOwnedShards())
```
Comment (notbdu) [inline src/dbnode/storage/bootstrap/bootstrapper/README.md:18]: Needs a callout here for peers bootstrapper i think - it takes INITIALIZING shards.

**Review by notbdu — APPROVED** [review]
Overall LGTM

**Comment by notbdu** [comment]
Can we also add these docs to the README as well w.r.t. to the shard states -> 
https://m3db.io/docs/m3db/architecture/sharding/#cluster-operations ? 

Probably can be included in the `Topology Changes` section.

File: src/dbnode/storage/bootstrap/bootstrapper/README.md (hunk 4 above)
Comment (vpranckaitis) [reply src/dbnode/storage/bootstrap/bootstrapper/README.md]: https://github.com/m3db/m3/pull/3203/commits/241e69f468515145c05fed2537891dad9bc29517

**Pull request merged**